core/store/src/migrations.rs

```diff
@@ -790,3 +794,70 @@ pub fn migrate_18_to_new_validator_stake(store: &Store) {
     }
     store_update.finish().unwrap();
 }
+
+pub fn fill_col_outcomes_with_metadata(store: &Store) {
+    #[derive(BorshSerialize, BorshDeserialize, PartialEq, Clone, Default, Eq, Debug)]
+    pub struct OldExecutionOutcome {
+        pub logs: Vec<LogEntry>,
+        pub receipt_ids: Vec<CryptoHash>,
+        pub gas_burnt: Gas,
+        pub tokens_burnt: Balance,
+        pub executor_id: AccountId,
+        pub status: ExecutionStatus,
+    }
+
+    #[derive(PartialEq, Clone, Default, Debug, BorshSerialize, BorshDeserialize, Eq)]
+    pub struct OldExecutionOutcomeWithId {
+        pub id: CryptoHash,
+        pub outcome: OldExecutionOutcome,
+    }
+
+    #[derive(PartialEq, Clone, Default, Debug, BorshSerialize, BorshDeserialize, Eq)]
+    pub struct OldExecutionOutcomeWithIdAndProof {
+        pub proof: MerklePath,
+        pub block_hash: CryptoHash,
+        pub outcome_with_id: OldExecutionOutcomeWithId,
+    }
+
+    impl Into<ExecutionOutcomeWithIdAndProof> for OldExecutionOutcomeWithIdAndProof {
+        fn into(self) -> ExecutionOutcomeWithIdAndProof {
+            ExecutionOutcomeWithIdAndProof {
+                proof: self.proof,
+                block_hash: self.block_hash,
+                outcome_with_id: ExecutionOutcomeWithId {
+                    id: self.outcome_with_id.id,
+                    outcome: ExecutionOutcome {
+                        logs: self.outcome_with_id.outcome.logs,
+                        receipt_ids: self.outcome_with_id.outcome.receipt_ids,
+                        gas_burnt: self.outcome_with_id.outcome.gas_burnt,
+                        tokens_burnt: self.outcome_with_id.outcome.tokens_burnt,
+                        executor_id: self.outcome_with_id.outcome.executor_id,
+                        status: self.outcome_with_id.outcome.status,
+                        metadata: ExecutionMetadata::ExecutionMetadataV1,
+                    },
+                },
+            }
+        }
+    }
+
+    let mut store_update = store.store_update();
+    let outcomes: Vec<(_, Vec<OldExecutionOutcomeWithIdAndProof>)> = store
+        .iter(DBCol::ColTransactionResult)
+        .map(|(key, value)| {
+            (
+                key,
+                Vec::<OldExecutionOutcomeWithIdAndProof>::try_from_slice(&value)
+                    .expect("BorshDeserialize should not fail"),
+            )
+        })
+        .collect();
```

Comment:
This loads all the outcomes to memory in the same time, is this OK to do? I'd expect that we either:

* iterate and update records in-place
* get the full key-set at once, and then iterate one by one


This might be premature optimization though, I'll let @bowenwang1996 to make a judgment call on whether we should bother with O(1) memory solution here.

Reply:
Good point

Reply:
We absolutely cannot load all of them in memory. In fact, I am not sure what the migration is for. Are we going to repopulate the old execution outcomes with metadata? If not, do we actually need to have this migration? Could we implement a custom deserializer instead?